File: PHP/Additional algorithm 38.php
<?php

	//Paul A. Gagniuc. An Introduction to Programming Languages: Simultaneous Learning in Multiple Coding Environments. Synthesis Lectures on Computer Science. Springer International Publishing, 2023, pp. 1-280.
	//Additional algorithm 38. It shows different experiments on recursive functions. A total of six examples are shown, in which: 1) A recursive function repeats one (or a group) of characters n times and returns a string of length n. 2) A recursive function sums integers from zero to n. 3) A recursive function computes the factorial for an integer n. 4) A function generated a sequence of numbers based on various rules. 5) A recursive function provides the Fibonacci sequence. 6) A recursive function sums all the integers stored in the elements of an array variable. Note that the source code is in context and works with copy/paste.

    // repeat string n times
    $a = x("*", "", 10);
    echo("Repeat:\n[" . $a . "]\n");
    
    // sum from 0 to n
    $b = sum(23);
    echo("Sum:[" . $b . "]\n");
    
    // factorial from 0 to n
    $c = factorial(10);
    echo("Factorial:\n[" . $c . "]\n");
    
    // just a sequence
    $s=""; $d = sequence(5,  [], 0, 5);
    foreach($d as $i){$s .= $i . ",";}
    echo("A sequence:\n[" . $s . "]\n");
    
    // fibonacci
    $s=""; $e = fibonacci(2, [0, 1, 2], 5);
    foreach($e as $i){$s .= $i . ",";}
    echo("Fibonacci:\n[" . $s . "]\n");
    
    // sum all from array
    $q = [1, 3, 3, 4, 5, 9];
    $f = sum_array(count($q) - 1, $q, 0);
    echo ("Sum array:[" . $f . "]");
    
    
    // repeat string n times
    function x($c, $s, $n){
        
        $s .= $c;
        
        if(strlen($s) >= $n){
            return $s;
        } else {
            return x($c, $s, $n);
        }
    }
    
    
    // sum from 0 to n
    function sum($n){
        if ($n <= 1) {return $n;}
        return $n + sum($n - 1);
    }
    
    
    // factorial from 0 to n
    function factorial($n){
        if ($n <= 1) {
            return $n;
        } else {
            return factorial($n - 1) * $n;
        }
    }
    
    
    // just a sequence
    function sequence($n, $m, $i, $t){
    
        $m[$i] = $n;
        $i++;
        
        if ($i >= $t) {
            return $m;
        } else {
        return sequence(($n-1)+($n-2), 
                            $m, $i, $t);
        }
    }
    
    
    // fibonacci
    function fibonacci($n, $m, $t){
        
        $n++;
        $m[$n] = $m[$n-1] + $m[$n-2];
        
        if ($n >= $t) {
            return $m;
        } else {
            return fibonacci($n, $m, $t);
        }
    }
    
    
    // sum all from array
    function sum_array($n, $q, $r){
        
        $r += $q[$n];
        
        if ($n <= 0) {
            return $r;
        } else {
            return sum_array($n - 1, $q, $r);
        }
    }

?>

File: Java/Additional algorithm 38.java
//Paul A. Gagniuc. An Introduction to Programming Languages: Simultaneous Learning in Multiple Coding Environments. Synthesis Lectures on Computer Science. Springer International Publishing, 2023, pp. 1-280.
//Additional algorithm 38. It shows different experiments on recursive functions. A total of six examples are shown, in which: 1) A recursive function repeats one (or a group) of characters n times and returns a string of length n. 2) A recursive function sums integers from zero to n. 3) A recursive function computes the factorial for an integer n. 4) A function generated a sequence of numbers based on various rules. 5) A recursive function provides the Fibonacci sequence. 6) A recursive function sums all the integers stored in the elements of an array variable. Note that the source code is in context and works with copy/paste.

public class Main
{
    public static void main(String[] args) {
        
    // repeat string n times
    String a = x("*", "", 10);
    System.out.println("Repeat:\n[" + 
                    a + "]"
                      );
    
    // sum from 0 to n
    int b = sum(23);
    System.out.println("Sum:[" + b + "]");
    
    // factorial from 0 to n
    int c = factorial(10);
    System.out.println(
          "Factorial:\n[" + c + "]"
                      );
    
    // just a sequence
    int[] d = sequence(5, new int[5], 0, 5);
    String l = "";
    for(int i = 0; i < d.length; i++){
        l += d[i] + ",";
        }

    System.out.println(
           "A sequence:\n[" + l + "]"
                      );
    
    // fibonacci
    l = "";
    int[] e = fibonacci(2, 
              new int[]{1,2,3,0,0,0}, 5);
    for(int i = 0; i < e.length; i++){
        l += e[i] + ",";
        }

    System.out.println(
            "Fibonacci:\n[" + l + "]"
                      );
    
    // sum all from array
    int[] q = {1, 3, 3, 4, 5, 9};
    int f = sum_array(q.length - 1, q, 0);
    System.out.println(
           "Sum array:[" + f + "]"
                      );
    }


    // repeat string n times
    static String x(String c, String s, 
                                  int n){
        s += c;
        
        if(s.length() >= n){
            return s;
        } else {
            return x(c, s, n);
        }
    }
    
    
    // sum from 0 to n
    static int sum(int n){
        if (n <= 1) {return n;}
        return n + sum(n - 1);
    }
    
    
    // factorial from 0 to n
    static int factorial(int n){
        if (n <= 1) {
            return n;
        } else {
            return factorial(n - 1) * n;
        }
    }
    
    
    // just a sequence
    static int[] sequence(int n, int[] m, 
                          int i, int t){
        m[i] = n;
        i++;
        
        if (i >= t) {
            return m;
        } else {
            return sequence((n-1)+(n-2), 
                               m, i, t);
        }
    }
    
    
  // fibonacci
  static int[] fibonacci(int n, int[] m, 
                         int t){
        n++;
        m[n] = m[n-1] + m[n-2];
        
        if (n >= t) {
            return m;
        } else {
            return fibonacci(n, m, t);
        }
    }

    
    // sum all from array
    static int sum_array(int n, int[] q, 
                         int r){
        r += q[n];
        
        if (n <= 0) {
            return r;
        } else {
            return sum_array(n - 1, q, r);
        }
    }
}
